roslibrust/src/ros1/service_client.rs

```diff
@@ -0,0 +1,172 @@
+use std::marker::PhantomData;
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::TcpStream,
+    sync::{
+        mpsc::{self, UnboundedReceiver},
+        oneshot,
+    },
+};
+
+pub type CallServiceRequest = (Vec<u8>, oneshot::Sender<CallServiceResponse>);
+pub type CallServiceResponse = RosLibRustResult<Vec<u8>>;
+
+pub struct ServiceClient<T: RosServiceType> {
+    service_name: Name,
+    sender: mpsc::UnboundedSender<CallServiceRequest>,
+    _phantom: PhantomData<T>,
+}
+
+impl<T: RosServiceType> ServiceClient<T> {
+    pub fn new(
+        service_name: &Name,
+        sender: mpsc::UnboundedSender<CallServiceRequest>,
+    ) -> ServiceClient<T> {
+        Self {
+            service_name: service_name.to_owned(),
+            sender,
+            _phantom: PhantomData,
+        }
+    }
+
+    pub fn service_name(&self) -> &Name {
+        &self.service_name
+    }
+
+    pub async fn call(&self, request: &T::Request) -> RosLibRustResult<T::Response> {
+        let request_payload = serde_rosmsg::to_vec(request)
+            .map_err(|err| RosLibRustError::SerializationError(err.to_string()))?;
+        let (response_tx, response_rx) = oneshot::channel();
+
+        self.sender
+            .send((request_payload, response_tx))
+            .map_err(|_err| RosLibRustError::Disconnected)?;
+
+        match response_rx.await {
+            Ok(Ok(result_payload)) => {
+                let response: T::Response = serde_rosmsg::from_slice(&result_payload[..])
+                    .map_err(|err| RosLibRustError::SerializationError(err.to_string()))?;
+                return Ok(response);
+            }
+            Ok(Err(err)) => {
+                return Err(err);
+            }
+            Err(_err) => {
+                return Err(RosLibRustError::Disconnected);
+            }
+        }
+    }
+}
+
+pub struct ServiceServerLink {
+    service_type: String,
+    call_sender: mpsc::UnboundedSender<CallServiceRequest>,
+    _actor_task: ChildTask<()>,
+}
+
+impl ServiceServerLink {
+    pub async fn new(
+        node_name: &Name,
+        service_name: &str,
+        service_type: &str,
+        service_uri: &str,
+        srv_definition: &str,
+        md5sum: &str,
+    ) -> RosLibRustResult<Self> {
+        let header = ConnectionHeader {
+            caller_id: node_name.to_string(),
+            latching: false,
+            msg_definition: srv_definition.to_owned(),
+            md5sum: md5sum.to_owned(),
+            topic: service_name.to_owned(),
+            topic_type: service_type.to_owned(),
+            tcp_nodelay: false,
+        };
+
+        let (call_tx, call_rx) = mpsc::unbounded_channel::<CallServiceRequest>();
+
+        let stream = establish_connection(&node_name, &service_name, &service_uri, header).await.map_err(|err| {
+            log::error!("Failed to establish connection to service URI {service_uri} for service {service_name}: {err}");
+            RosLibRustError::from(err)
+        })?;
```

Comment:
Ultimately this is just coercing the `std::io::Error` into a `RosLibRustError::IoError`. In the future I'll update `establish_connection` to do this (when I update the publisher actor to use it as well).